rfc-143-split-database-instances.md

```diff
@@ -0,0 +1,66 @@
+
+Applying this initially to our databases, we propose that:
+
+1. All new long-lived databases are created in their own individual RDS instances.
+2. Existing MySQL and Postgres databases are migrated out of the central instance to individual managed-service instances as part of the upcoming major upgrade cycle.
+3. Existing Redis and Mongo databases are migrated out of their central instances in-line with team priorities, and prior to any major upgrade.
+4. Any future supporting services for applications should be distinct to the application and, where applicable, run on a managed service
+
+
+### Consequences
+
+- It will be easier to identify the causes of performance problems in an application.
+
+- The chance of a problem with one service cascading to other services will be reduced, because there is no contention for resources which aren't shared.
+
+- Creating more RDS instances means that application-owning teams will need to monitor and consider database upgrades as part of their day-to-day work. We can ease this burden through monitoring & reporting, which may be considered by a future Platform Reliability team.
+
+- There will be less consistency in database versions across the programme. Individual teams are responsible for their upgrade appetite and prioritisation.
+
+- There is a potential increase in costs caused by having dedicated instances, particularly when they are initially split (because it's harded to accurately estimate resource requirements). This may be offset by being able to optimise individual databases to their needs.
+
+- Migrating to new instances with minimal (or no) downtime takes planning and effort. AWS' forced upgrades of Postgres and MySQL in early 2022 will force this effort regardless.
+
+- This change affects our production, integration, staging and CI environments. Our setup for CI requires a slightly different user setup so we can create and destroy databases within the instance (where we create databases per test run).
+
+- CI will need to support a wider range of database versions than it currently supports to cope with teams using different versions. Ideally, teams will be able to create and manage databases to meet their needs independently, but as a first step we'll need to introduce a pool of "supported" database types and versions. We have experience of doing this with Mongo and tagged agents.
+
+## Questions
+
+1. Can we be specific about databases. If we increase scope to all supporting services, does this RFC become less actionable now?
```

Comment:
It adds a bit of complexity, yes - I wouldn't be against removing supporting services from the scope, and introducing that as a follow-up RFC once the DBs are migrated.

Reply:
🙌 this question was left over from an initial draft. I _think_ we've got a reasonable balance that lets us be specific about databases while also saying it applies to supporting services now.